core/src/mast/mod.rs

```diff
@@ -393,6 +466,40 @@ impl MastForest {
     pub fn nodes(&self) -> &[MastNode] {
         &self.nodes
     }
+
+    /// Returns an iterator which traverses over the nodes in a depth-first search and returns nodes
+    /// in postorder.
+    ///
+    /// This iterator iterates through all **reachable** nodes of a forest exactly once.
+    ///
+    /// Since a `MastForest` has multiple possible entrypoints in the form of its roots, a
+    /// depth-first search must visit all of those roots and the trees they form.
+    ///
+    /// For instance, consider this `MastForest`:
+    ///
+    /// ```text
+    /// Nodes: [Block(foo), Block(bar), Join(0, 1), External(qux)]
+    /// Roots: [2]
+    /// ```
+    ///
+    /// The only root is the `Join` node at index 2. The first three nodes of the forest form a
+    /// tree, since the `Join` node references index 0 and 1. This tree is discovered by
+    /// starting at the root at index 2 and following all children until we reach terminal nodes
+    /// (like `Block`s) and build up a stack of the discovered, but unvisited nodes. The stack
+    /// is built such that popping elements off the stack (from the back) yields a postorder.
+    ///
+    /// After the first tree is discovered, the stack looks like this: `[2, 1, 0]`. On each
+    /// call to `next` one element is popped off this stack and returned.
+    ///
+    /// If the stack is exhausted we start another discovery if more unvisited roots exist. Since
+    /// the `External` node is not a root and not referenced by any other tree in the forest, it
+    /// will not be visited.
+    ///
+    /// The iteration on a high-level thus consists of a constant back and forth between discovering
+    /// trees and returning nodes from the stack.
+    pub fn iter_nodes(self) -> impl Iterator<Item = (MastNodeId, MastNode)> {
+        MastForestNodeIter::new(self)
+    }
```

Comment:
Related to the above comment: does `iter_nodes()` need to take ownership of `self`? Maybe we could pass in just `&self`?

Reply:
If `MastForestMerger::merge()` is going to take ownership of the other forest (as discussed in previous comment), then this needs to take ownership so that we can avoid clones.

nit: I would then call it `into_iter_nodes()`